### ISSUE TITLE
tests.test_trees.TestTrees.test_smart_decorator fails with python 3.11

### ISSUE THREAD

**Issue by mgorny** [lark-parser/lark#1146]
**Describe the bug**

When running the test suite (without js2py installed, see https://github.com/PiotrDabkowski/Js2Py/issues/282), the following test fails:

```pytb
======================================================================
ERROR: test_smart_decorator (tests.test_trees.TestTrees.test_smart_decorator)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/lark/lark/visitors.py", line 114, in _call_userfunc
    return f.visit_wrapper(f, tree.data, children, tree.meta)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lark/lark/visitors.py", line 491, in _vargs_inline
    return f(*children)
           ^^^^^^^^^^^^
  File "/tmp/lark/lark/visitors.py", line 471, in __call__
    return self.base_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lark/tests/test_trees.py", line 191, in ab_classmethod
    assert cls is OtherClass, cls
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: <tests.test_trees.TestTrees.test_smart_decorator.<locals>.TestCls object at 0x7f5bcf0c7390>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/lark/tests/test_trees.py", line 293, in test_smart_decorator
    result = test_instance.transform(Tree(method_name, children))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lark/lark/visitors.py", line 153, in transform
    return self._transform_tree(tree)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lark/lark/visitors.py", line 149, in _transform_tree
    return self._call_userfunc(tree, children)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lark/lark/visitors.py", line 120, in _call_userfunc
    raise VisitError(tree.data, tree, e)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lark.exceptions.VisitError: Error trying to process rule "oc_ab_classmethod":

<tests.test_trees.TestTrees.test_smart_decorator.<locals>.TestCls object at 0x7f5bcf0c7390>
```

**To Reproduce**

```
python3.11 -m venv .venv
. .venv/bin/activate
python -m tests
```

**Comment by MegaIng** [comment]
Yeah, this is not fully unexpected. The smart decorators depend on a few semi-implementation details. I will see if I can get this fixed later.

**Comment by MegaIng** [comment]
Ok, I am failing to understand what change happened to break this. The core of the problem is that the beahvior of this changed:

```py

class A:
    @classmethod
    def test(cls):
        print(f"A.test got called with {cls=}")

class B:
    A_test = A.test
     
B.A_test()
A.test.__get__(B, B)()
```
Previously this would in both cases call with `cls=A`, but now in py3.11 in the bottom case it calls with `cls=B`, and I don't know why the two statements are different. AFAIK they should be exactly equivalent. It might be related to https://github.com/python/cpython/issues/89519 , but I don't fully understand what changes happened there.

**Comment by MegaIng** [comment]
Ok, figured it out. It was a mistake on my part when originally implementing this behavior, that got exposed by https://github.com/python/cpython/pull/31367 . Fixed in #1147 .

The problem is that I was doing `A.test.__get__` instead of `type(A.test).__get__`, which changed behavior since in 3.11 `type(A.test).__get__` no longer exists, so the former would fallback to the `__get__` of the function, which is incorrect.

**Comment by mgorny** [comment]
Thanks. I can confirm that the newest release passes all tests.
